pkg/operator/ceph/cluster/osd/spec.go

```diff
@@ -74,23 +74,36 @@ const (
 
 const (
 	cephOSDStart = `
-function sigterm() {
-	echo "SIGTERM received"
+RESTART_INTERVAL=%d
+child_pid=""
+
+function handle_signal() {
+	echo "Signal received"
+	# kill the child process is its still running
+	kill -TERM "$child_pid"
 	exit
 }
 
-trap sigterm SIGTERM
+trap handle_signal SIGTERM
 
-%s %s & wait
+%s %s &
+child_pid="$!"
+wait "$child_pid"
+# osd_rc stores the exit code from the wait command.
+# Wait command returns the exit code of the child process.
+# If process is killed by a signal then wait returns a non zero exit code of that signal.
+osd_rc=$?
+echo "osd process exit code $osd_rc"
 
-RESTART_INTERVAL=%d
-rc=$?
-if [ $rc -eq 0 ]; then
+if [ $osd_rc -eq 0 ]; then
 	touch /tmp/osd-sleep
 	echo "OSD daemon exited with code 0, possibly due to OSD flapping. The OSD pod will sleep for $RESTART_INTERVAL hours. Restart the pod manually once the flapping issue is fixed"
-	sleep "$RESTART_INTERVAL"h & wait
-	exit $rc
-fi`
+	sleep "$RESTART_INTERVAL"h &
+	child_pid="$!"
+	wait "$child_pid"
+fi
+
+exit $osd_rc`
```

Comment:
Try this. The current script didn't account for the osd returning 0 when it received sigterm from bash.

```sh
RESTART_INTERVAL=%d

child_pid=""
sigterm_received=false

function sigterm() {
  echo "SIGTERM received"
  sigterm_received=true
  kill -TERM "$child_pid"
}
trap sigterm SIGTERM

bash -c "sleep 5 && false" &
# un-fixable race condition: if receive sigterm here, it won't be sent to child process
child_pid="$!"
wait "$child_pid" # wait returns the same return code of child process when called with argument
wait "$child_pid" # first wait returns immediately upon SIGTERM, so wait again for child to actually stop; this is a noop if child exited normally
ceph_osd_rc=$?
if [ $ceph_osd_rc -eq 0 ] && ! $sigterm_received; then
  touch /tmp/osd-sleep
  echo "OSD daemon exited with code 0, possibly due to OSD flapping. The OSD pod will sleep for $RESTART_INTERVAL hours. Restart the pod manually once the flapping issue is fixed"
  sleep "$RESTART_INTERVAL"h &
  child_pid="$!"
  wait "$child_pid"
  wait "$child_pid" # wait again for sleep to stop
fi
exit $ceph_osd_rc
```

Reply:
> Try this. The current script didn't account for the osd returning 0 when it received sigterm from bash.
> 
> ```shell
> RESTART_INTERVAL=%d
> 
> child_pid=""
> sigterm_received=false
> 
> function sigterm() {
>   echo "SIGTERM received"
>   sigterm_received=true
>   kill -TERM "$child_pid"
> }
> trap sigterm SIGTERM
> 
> bash -c "sleep 5 && false" &
> # un-fixable race condition: if receive sigterm here, it won't be sent to child process
> child_pid="$!"
> wait "$child_pid" # wait returns the same return code of child process when called with argument
> wait "$child_pid" # first wait returns immediately upon SIGTERM, so wait again for child to actually stop; this is a noop if child exited normally

Here if the child exits normally the second `wait` fails to wait for the child pid and returns 127 exit code.  So `ceph_osd_rc` becomes `127`

> ceph_osd_rc=$?
> if [ $ceph_osd_rc -eq 0 ] && ! $sigterm_received; then
>   touch /tmp/osd-sleep
>   echo "OSD daemon exited with code 0, possibly due to OSD flapping. The OSD pod will sleep for $RESTART_INTERVAL hours. Restart the pod manually once the flapping issue is fixed"
>   sleep "$RESTART_INTERVAL"h &
>   child_pid="$!"
>   wait "$child_pid"
>   wait "$child_pid" # wait again for sleep to stop
> fi
> exit $ceph_osd_rc
> ```